src/core/module.cpp

```diff
@@ -84,8 +84,9 @@ void Module::LoadModuleToMemory(u32& max_tls_index) {
 
     // Map module segments (and possible TLS trampolines)
     void** out_addr = reinterpret_cast<void**>(&base_virtual_addr);
-    memory->MapMemory(out_addr, LoadAddress, aligned_base_size + TrampolineSize,
-                      MemoryProt::CpuReadWrite, MemoryMapFlags::Fixed, VMAType::Code, name, true);
+    memory->MapMemory(out_addr, memory->VirtualOffset() + LoadAddress,
```

Comment:
Why is this needed? If the virtual base is allocated too high it will break a lot of games

Reply:
I haven't been able to find any way to reserve the range that is currently asked for; the best I've been able to get is starting at `0x7000000000`. The reason being that dynamic libraries and Rosetta runtime components may be allocated in the range first.

There's a workaround that wine uses to reserve low address space by creating a `zerofill` section in the executable covering the desired range, but the size of the virtual memory space is way too large, and creating a `zerofill` section that large causes it to crash on load.

Reply:
Is there a known limit on how high the virtual base can be before games start to break?

Reply:
PS4 memory system has system managed area, system reserved and user area. There are games that expect mappings to be at most 40bits in size, so higher addresses are not allowed. For system managed the emulator currently reserves from the lowest portion of the area, however we can get away with starting from 0xfc0000000 instead as red-prig discovered in kernel that is the search start.
So I would say you can try the zerofill trick from 0xfc0000000 to end of system managed (there are constants for these in address_space.h). Then let Rosetta take whatever it wants from system reserved area (hopefully it doesn't reach user area) and reserve user area.

Reply:
There seems to be a region starting somewhere near 0x100000000 and ending at 0x7000000000 that just refuses to be mapped, even if looking at the virtual memory map of the process it appears to be unused. I've tried tricks like reserving the area for stack or other memory in the executable, and while that works for regions higher up in memory, if it overlaps with that unmappable region my system actually crashes completely. It turns out that this is the issue I was having earlier, not that the zerofill was too big.

So I don't think it's actually possible to get the user area mapping any lower than 0x7000000000

Reply:
User area usage is less common as system managed so as long as that is reserved properly it should be a good first step compatibility wise

Reply:
Is it okay if the system managed and user space areas are discontiguous? I think I could maybe get the system area reserved going up to its end, but the user area would have to go all the way over at 0x7000000000 as mentioned.